File: src/main/java/br/com/shoppinglist/shopping_list/identity/application/usecase/ConfirmRegistrationUseCaseImpl.java
package br.com.shoppinglist.shopping_list.identity.application.usecase;

import br.com.shoppinglist.shopping_list.identity.application.dto.AuthTokenResult;
import br.com.shoppinglist.shopping_list.identity.application.dto.ConfirmRegistrationCommand;
import br.com.shoppinglist.shopping_list.identity.application.dto.EmailVerificationCode;
import br.com.shoppinglist.shopping_list.identity.application.port.in.ConfirmRegistrationUseCase;
import br.com.shoppinglist.shopping_list.identity.application.port.out.EmailVerificationCodeRepository;
import br.com.shoppinglist.shopping_list.identity.application.port.out.PasswordHasher;
import br.com.shoppinglist.shopping_list.identity.application.port.out.TokenIssuer;
import br.com.shoppinglist.shopping_list.identity.application.port.out.UserRepository;
import br.com.shoppinglist.shopping_list.identity.domain.exception.EmailAlreadyInUseException;
import br.com.shoppinglist.shopping_list.identity.domain.exception.InvalidVerificationCodeException;
import br.com.shoppinglist.shopping_list.identity.domain.exception.VerificationCodeExpiredException;
import br.com.shoppinglist.shopping_list.identity.domain.model.User;
import java.time.Instant;
import org.springframework.stereotype.Service;

@Service
public class ConfirmRegistrationUseCaseImpl implements ConfirmRegistrationUseCase {

	private final EmailVerificationCodeRepository emailVerificationCodeRepository;
	private final UserRepository userRepository;
	private final PasswordHasher passwordHasher;
	private final TokenIssuer tokenIssuer;

	public ConfirmRegistrationUseCaseImpl(
		EmailVerificationCodeRepository emailVerificationCodeRepository,
		UserRepository userRepository,
		PasswordHasher passwordHasher,
		TokenIssuer tokenIssuer
	) {
		this.emailVerificationCodeRepository = emailVerificationCodeRepository;
		this.userRepository = userRepository;
		this.passwordHasher = passwordHasher;
		this.tokenIssuer = tokenIssuer;
	}

	@Override
	public AuthTokenResult execute(ConfirmRegistrationCommand command) {
		String email = RegistrationInputValidator.normalizeEmail(command.email());
		String code = normalizeCode(command.code());
		RegistrationInputValidator.validate(email, command.name(), command.password());

		if (code.isBlank()) {
			throw new InvalidVerificationCodeException();
		}

		EmailVerificationCode verificationCode = emailVerificationCodeRepository.findByEmail(email)
			.orElseThrow(InvalidVerificationCodeException::new);

		if (verificationCode.expiresAt().isBefore(Instant.now())) {
			emailVerificationCodeRepository.deleteByEmail(email);
			throw new VerificationCodeExpiredException();
		}

		if (!passwordHasher.matches(code, verificationCode.codeHash())) {
			throw new InvalidVerificationCodeException();
		}

		if (userRepository.existsByEmail(email)) {
			emailVerificationCodeRepository.deleteByEmail(email);
			throw new EmailAlreadyInUseException(email);
		}

		String passwordHash = passwordHasher.hash(command.password());
		User user = userRepository.save(email, command.name().trim(), passwordHash);
		emailVerificationCodeRepository.deleteByEmail(email);
		return tokenIssuer.issue(user.id());
	}

	private String normalizeCode(String code) {
		if (code == null) {
			return "";
		}
		return code.trim();
	}
}

File: src/identity/application/usecase/confirm-registration.use-case.impl.ts
import { Injectable } from '@nestjs/common';
import {
  EmailAlreadyInUseException,
  InvalidVerificationCodeException,
  VerificationCodeExpiredException,
} from '../../domain/exception/identity.exceptions';
import { AuthTokenResult } from '../dto/auth-token.result';
import { ConfirmRegistrationCommand } from '../dto/confirm-registration.command';
import { ConfirmRegistrationUseCase } from '../port/in/confirm-registration.use-case';
import { EmailVerificationCodeRepository } from '../port/out/email-verification-code.repository';
import { PasswordHasher } from '../port/out/password-hasher';
import { TokenIssuer } from '../port/out/token-issuer';
import { UserRepository } from '../port/out/user.repository';
import { RegistrationInputValidator } from './registration-input.validator';

@Injectable()
export class ConfirmRegistrationUseCaseImpl extends ConfirmRegistrationUseCase {
  constructor(
    private readonly verificationCodes: EmailVerificationCodeRepository,
    private readonly users: UserRepository,
    private readonly passwordHasher: PasswordHasher,
    private readonly tokenIssuer: TokenIssuer,
  ) {
    super();
  }

  async execute(command: ConfirmRegistrationCommand): Promise<AuthTokenResult> {
    const email = RegistrationInputValidator.normalizeEmail(command.email);
    const code = (command.code ?? '').trim();
    RegistrationInputValidator.validate(email, command.name, command.password);

    if (!code) {
      throw new InvalidVerificationCodeException();
    }

    const verification = await this.verificationCodes.findByEmail(email);
    if (!verification) {
      throw new InvalidVerificationCodeException();
    }

    if (verification.expiresAt.getTime() < Date.now()) {
      await this.verificationCodes.deleteByEmail(email);
      throw new VerificationCodeExpiredException();
    }

    if (!(await this.passwordHasher.matches(code, verification.codeHash))) {
      throw new InvalidVerificationCodeException();
    }

    if (await this.users.existsByEmail(email)) {
      await this.verificationCodes.deleteByEmail(email);
      throw new EmailAlreadyInUseException(email);
    }

    const user = await this.users.create(
      email,
      command.name.trim(),
      await this.passwordHasher.hash(command.password),
    );
    await this.verificationCodes.deleteByEmail(email);
    return this.tokenIssuer.issue(user.id);
  }
}
